ghcide/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -44,26 +56,53 @@ type IdeResult v = ([FileDiagnostic], Maybe v)
 -- | an IdeResult with a fingerprint
 type IdeResultNoDiagnosticsEarlyCutoff  v = (Maybe ByteString, Maybe v)
 
-ideErrorText :: NormalizedFilePath -> T.Text -> FileDiagnostic
-ideErrorText = ideErrorWithSource (Just "compiler") (Just DiagnosticSeverity_Error)
+ideErrorText :: Maybe (MsgEnvelope GhcMessage) -> NormalizedFilePath -> T.Text -> FileDiagnostic
+ideErrorText origMsg fdFilePath msg =
+  ideErrorWithSource (Just "compiler") (Just DiagnosticSeverity_Error) fdFilePath msg origMsg
+
+ideErrorFromLspDiag
+  :: LSP.Diagnostic
+  -> NormalizedFilePath
+  -> Maybe (MsgEnvelope GhcMessage)
+  -> FileDiagnostic
+ideErrorFromLspDiag lspDiag fdFilePath origMsg =
+  let fdShouldShowDiagnostic = ShowDiag
+      fdStructuredMessage =
+        maybe NoStructuredMessage SomeStructuredMessage origMsg
+      fdLspDiagnostic = lspDiag
+        { _code = fmap ghcCodeToLspCode . diagnosticCode . errMsgDiagnostic =<< origMsg
+        }
+      ghcCodeToLspCode :: DiagnosticCode -> Int32 LSP.|? T.Text
+#if MIN_VERSION_ghc(9,10,1)
+      ghcCodeToLspCode = InR . T.pack . show
+#else
+      ghcCodeToLspCode (DiagnosticCode prefix c) = InR $ T.pack $ prefix ++ "-" ++ printf "%05d" c
+#endif
+  in
+  FileDiagnostic {..}
 
 ideErrorWithSource
   :: Maybe T.Text
   -> Maybe DiagnosticSeverity
-  -> a
+  -> NormalizedFilePath
   -> T.Text
-  -> (a, ShowDiagnostic, Diagnostic)
-ideErrorWithSource source sev fp msg = (fp, ShowDiag, LSP.Diagnostic {
-    _range = noRange,
-    _severity = sev,
-    _code = Nothing,
-    _source = source,
-    _message = msg,
-    _relatedInformation = Nothing,
-    _tags = Nothing,
-    _codeDescription = Nothing,
-    _data_ = Nothing
-    })
+  -> Maybe (MsgEnvelope GhcMessage)
+  -> FileDiagnostic
+ideErrorWithSource source sev fdFilePath msg origMsg =
+  let lspDiagnostic =
+        LSP.Diagnostic {
+          _range = noRange,
+          _severity = sev,
+          _code = Nothing,
```

Comment:
set this I guess

Reply:
It gets set via `ideErrorFromLspDiag`, which is the underlying function that `ideErrorWithSource` calls out to.

I agree that the code path here is a bit strange, but I can't really decide on a better separation of concerns.